modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/snapshot/MajorityPeerSelect.scala

```diff
@@ -50,28 +59,34 @@ object MajorityPeerSelect {
 
     def filterPeerList(peers: NonEmptyList[Peer]): F[NonEmptyList[Peer]] =
       peers
-        .traverse(snapshotClient.getLatestOrdinal(_))
-        .map {
-          _.groupBy(identity).maxBy { case (_, ordinals) => ordinals.size }
+        .parTraverseN(maxConcurrentPeerInquiries)(snapshotClient.getLatestOrdinal(_))
+        .map(_.groupBy(identity).maxBy { case (_, ordinals) => ordinals.size })
+        .map { case (majorityOrdinal, _) => majorityOrdinal }
+        .flatMap { majorityOrdinal =>
+          peers
+            .parTraverseN(maxConcurrentPeerInquiries)(peer => getPeerHash(peer, majorityOrdinal).attempt)
         }
-        .flatMap {
-          case (majorityOrdinal, _) =>
-            peers.traverse(snapshotClient.get(majorityOrdinal).run(_).flatMap(_.toHashed.map(_.hash)))
+        .flatMap { maybePeerSnapshots =>
+          MonadThrow[F].fromOption(
+            maybePeerSnapshots.collect { case Right(peerSnapshot) => peerSnapshot }.toNel,
+            NoPeersToSelect
+          )
         }
-        .map(_.zip(peers))
-        .map(_.groupMap { case (hash, _) => hash } { case (_, ps) => ps })
-        .map(_.maxBy { case (_, peers) => peers.size })
-        .map { case (_, peerCandidates) => peerCandidates }
+        .map(_.groupMap { case (_, hash) => hash } { case (peer, _) => peer })
+        .map(_.values.toList.sortWith(_.size > _.size))
+        .map(_.head)
 
     def getPeerSublist(peers: Set[Peer]): F[List[Peer]] = {
-      val maxSublistPercent = 0.25
-      val maxSublistSize = Math.max((peers.size * maxSublistPercent).toInt, 1)
+      val minSampleSize = Math.max((peers.size * maxPeerSampleRatio).toInt, 20)
```

Comment:
is it min sample size or max sublist size? I thought that `20` is the min sample size.

imo it should look like this:
```scala
val sublistPercent = 0.25
val minSampleSize = 20
val sublistSize = Math.max((peers.size * maxSublistPercent).toInt, minSampleSize)

Random[F].nextIntBounded(sublistSize).map(Math.max(_, 1)).flatMap { peersToTake =>
  Random[F]
    .shuffleList(peers.toList)
    .map(_.take(peersToTake))
}
```

Reply:
also, I'm thinking if `Random[F].nextIntBounded` is indeed correct 🤔 do you want to take an arbitrary peers count from 1 to 25% of all peers?

Reply:
Changed the implementation to the following:

```
      val maxSampleSize = minPeerSelectionSize + (peers.size * peerSampleRatio).toInt

      Random[F].betweenInt(minPeerSelectionSize, maxSampleSize + 1).flatMap { peerCount =>
        Random[F]
          .shuffleList(peers.toList)
          .map(_.take(peerCount))
      }
```

where

```
  val peerSampleRatio = 0.25
  val minPeerSelectionSize = 20
```

So, at least 20 `Peer`s will be selected. If `peer.size < 20` then `_.take(peerCount)` will select all the `Peer`s. The upper bound for the number of `Peer`s selected is `maxSampleSize` or 20 `Peer`s more than 25% of the total number of discovered `Peer`s. At low number of discovered `Peer`s, this means that most or all of the discovered `Peer`s will be selected. As the number of discovered `Peer`s grows, the `20` becomes increasingly dwarfed by `peers.size * peerSampleRatio` and increasingly trivial. Since the point of selecting a subset of `Peer`s is avoid to many queries as the number of discovered `Peer`s grows, this is fine.